### PR TITLE
translate module

### PR DESCRIPTION
**Pull request by robatipoor**

### CODE REVIEW

File: src/translate/mod.rs [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+impl Translate for str {}
+
+#[cfg(test)]
+mod translate {
+    use super::{Language, Translate};
+
+    #[test]
+    fn translate() {
+        assert_eq!(
+            "سلام دنیا",
```
Comment (kianenigma) [inline src/translate/mod.rs:125]: 👍🏻 excellent test :D

File: src/translate/mod.rs [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+                } else {
+                    Err("Does Not Exist".to_string())
+                }
+            })
+            .map_err(|e| e.into())
+    }
+}
+
+impl Translate for String {}
+
```
Comment (kianenigma) [inline src/translate/mod.rs:115]:
```suggestion
```

File: src/translate/mod.rs [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+            Language::Japanese => "ja",
+            Language::Portuguese => "pt",
+            Language::Russian => "ru",
+            Language::Spanish => "es",
+        }
+    }
+
+    /// Create a Language from &str like "en" or "French". Case Doesn't matter.
+    pub fn from(s: &str) -> crate::Result<Self> {
+        return Self::from_str(s).map_err(|e| e.into());
```
Comment (kianenigma) [inline src/translate/mod.rs:47]:
```suggestion
        Self::from_str(s).map_err(|e| e.into())
```

(where is this used? why not use FromStr directly?)

File: src/translate/mod.rs [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+    }
+}
+
+impl FromStr for Language {
+    type Err = &'static str;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        match s.to_string().to_lowercase().as_str() {
+            "en" => Ok(Language::English),
```
Comment (omid) [inline src/translate/mod.rs:56]: You can use pipeline here.
For example:
`"en"|"english" => ...`

File: src/translate/mod.rs [hunk 5]
```diff
@@ -0,0 +1,129 @@
...
+            "arabic" => Ok(Language::Arabic),
+            "chinese" => Ok(Language::Chinese),
+            "french" => Ok(Language::French),
+            "german" => Ok(Language::German),
+            "italian" => Ok(Language::Italian),
+            "portuguese" => Ok(Language::Portuguese),
+            "russian" => Ok(Language::Russian),
+            "spanish" => Ok(Language::Spanish),
+            "japanese" => Ok(Language::Japanese),
+            &_ => Err("input invalid!"),
```
Comment (omid) [inline src/translate/mod.rs:78]: This `&` is not required

File: src/translate/mod.rs [hunk 6]
```diff
@@ -0,0 +1,129 @@
...
+        }
+    }
+}
+
+pub trait Translate: AsRef<str> {
+    const BASE_URL: &'static str =
+        "https://translate.google.com/translate_a/single?&client=gtx&sl=auto";
+    const USER_AGENT_VALUE:&'static str = "Mozilla/5.0 (Windows NT 6.1; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/58.0.3029.110 Safari/537.36";
```
Comment (omid) [inline src/translate/mod.rs:86]: code style is not correct, please run `cargo fmt` or better, `make check`

**Comment by robatipoor** [comment]
Thanks guys

**Pull request merged**